iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
> EDIT: I suppose it's possible that you're suggesting two equal values may not be in the domain of ==, which would invalidate my argument, but leave me just as baffled.

This is what I've been arguing. Undoubtedly you will get *something* if you compare iterators from different vectors, but you'd be crazy to to anything with the result. Consider also the Range TS's `counted_iterator`, which only compares the internal count. Two counted iterators into different ranges may well compare equal (if their counts are the same), but that doesn't mean they point to the same element! I don't have to tell you that -- you know it already. Sentinels are no different, IMO.

As another example, think of default-constructed `istream_iterator<string>`s (the current STL's equivalent of a sentinel). They all compare equal. That is consistent with their usage: to denote the end of a range. It would be wrong-headed to try to also compare the state of the internal `string` data member!

Given some range `[first,last)` where `last` is a sentinel, and two iterators `i` and `j` that are reachable from `first`, then `i==last && j==last` implies `i==j`. *That* is the essential property of sentinels. And those semantics come straight from the algorithms and how they use iterators, sentinels, and `==`.

So I don't see a need to give up the notion that a sentinel denotes a position or that `==` means equality. `==` over sentinels means equality *in the sense in which `==` and sentinels are used by the algorithms*.